R/createCompoundDbPackage.R

```diff
@@ -0,0 +1,149 @@
+#' @title Create a CompoundDb database
+#'
+#' @description
+#' 
+#' `createCompoundDb` creates a `SQLite`-based `CompoundDb` object/database
+#' from a compound resource provided as a `data.frame` or `tbl`.
+#' An additional `data.frame` providing metadata information is mandatory.
+#' Required columns for the `data.frame` providing the compound information are:
+#'
+#' 
+#' The metadata `data.frame` is supposed to have two columns named `"name"` and
+#' `"value"` providing the following minimal information as key-value pairs:
+#' + `"source"`: the source from which the data was retrieved (e.g. `"HMDB"`).
+#' + `"url"`: the url from which the original data was retrieved.
+#' + `"source_version"`: the version from the original data source
+#'   (e.g. `"v4"`).
+#' + `"source_date"`: the date when the original data source was generated.
+#' + `"organism"`: the organism. Should be in the form `"Hsapiens"` or
+#'   `"Mmusculus"`.
+#'
+#' @details
+#'
+#' Metadata information is also used to create the file name for the database
+#' file. The name starts with `"CompoundDb"`, followed by the organism, the
+#' data source and its version. A compound database file for HMDB version 4
+#' with human metabolites will thus be named: `"CompoundDb.Hsapiens.HMDB.v4"`.
+#' 
+#' @param x `data.frame` or `tbl` with the compound annotations. See
+#'     description for details.
+#'
+#' @param metadata `data.frame` with metadata information. See description for
+#'     details.
+#'
+#' @param path `character(1)` with the path to the directory where the database
+#'     file should be written. Defaults to the current directory.
+#'
+#' @return A `character` with the database name (invisibly).
+#' 
+#' @importFrom DBI dbDriver dbWriteTable dbExecute
+#' @importFrom RSQLite dbConnect
+#'
+#' @export
+#'
+#' @author Johannes Rainer
+#'
+#' @md
+#'
+#' @examples
+#'
+#' ## Read compounds for a HMDB subset
+#' fl <- system.file("extdata/hmdb/hmdb_sub.xml", package = "PeakABro")
+#' cmps <- generate_hmdb_tbl(fl)
+#'
+#' ## Create a metadata data.frame for the compounds.
+#' metad <- data.frame(name = c("source", "url", "source_version",
+#'     "source_date", "organism"), value = c("HMDB", "http://www.hmdb.ca",
+#'     "v4", "2017-08-27", "Hsapiens"))
+#'
+#' ## Create a SQLite database in the temporary folder
+#' db_f <- createCompoundDb(cmps, metadata = metad, path = tempdir())
+#'
+#' ## connect to the database and query it's tables using RSQlite
+#' library(RSQLite)
+#' con <- dbConnect(dbDriver("SQLite"), db_f)
+#'
+#' dbGetQuery(con, "select * from metadata")
+#' dbGetQuery(con, "select * from compound")
+createCompoundDb <- function(x, metadata, path = ".") {
+    .valid_metadata(metadata)
+    .valid_compound(x)
+    db_file <- paste0(path, .db_file_from_metadata(metadata), ".sqlite")
+    con <- dbConnect(dbDriver("SQLite"), dbname = db_file)
```

Comment:
more consistent with tidyverse to use `src_sqlite`?
Instead of SQL select you can also use `dplyr::select`.

Example from the pubchem I am working on:
```r
pubchem_db <- src_sqlite(path=sqlite_path, create = TRUE) # create or connect
db_insert_into(pubchem_db$con, table="pubchem_tbl", values=pubchem_tbl)  #insert rows

pubchem_all <- tbl(pubchem_db$con, "pubchem_tbl") %>% collect() # read into tibble

rm(pubchem_db) # release the file
gc() # release the file
```

Reply:
uh, tried to use the `src_sqlite`, but then I have problems sending SQL queries to the connection (e.g. when I create the indices on certain database columns or count distinct elements etc). I'll need to do that a lot and prefer to have direct control over the query to send.

Reply:
Had some thoughts about internal tidyverse. For these, let's say, low level stuff I would like to stick to the low level packages (`DBI`) with the `dbConnect`, `dbGetQuery`, `dbWriteTable` etc because they do their stuff, and the `dplyr` (or `dbplyr`?) essentially just wraps them up. For examples, vignettes etc I'm OK with using tidyverse, but then we should consider to add `dplyr` to `DEPENDS` not only `IMPORTS`.

Reply:
OK. I am fine with that.